src/network/networkd-dhcp-common.c

```diff
@@ -445,27 +476,49 @@ int config_parse_dhcp_send_option(
                 return -EINVAL;
         }
 
-        r = sd_dhcp_option_new(u, udata, sz, &opt);
-        if (r < 0) {
-                log_syntax(unit, LOG_ERR, filename, line, r,
-                           "Failed to store DHCPv4 option '%s', ignoring assignment: %m", rvalue);
-                return 0;
-        }
+        if (dhcpv6) {
+                r = sd_dhcp6_option_new(u16, udata, sz, &opt6);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to store DHCP option '%s', ignoring assignment: %m", rvalue);
+                        return 0;
+                }
 
-        r = ordered_hashmap_ensure_allocated(options, &dhcp_option_hash_ops);
-        if (r < 0)
-                return log_oom();
+                r = ordered_hashmap_ensure_allocated(options, &dhcp6_option_hash_ops);
+                if (r < 0)
+                        return log_oom();
 
-        /* Overwrite existing option */
-        old = ordered_hashmap_remove(*options, UINT_TO_PTR(u));
-        r = ordered_hashmap_put(*options, UINT_TO_PTR(u), opt);
-        if (r < 0) {
-                log_syntax(unit, LOG_ERR, filename, line, r,
-                           "Failed to store DHCPv4 option '%s', ignoring assignment: %m", rvalue);
-                return 0;
-        }
+                /* Overwrite existing option */
+                old6 = ordered_hashmap_remove(*options, UINT_TO_PTR(u16));
+                r = ordered_hashmap_put(*options, UINT_TO_PTR(u16), opt6);
```

Comment:
ordered_hashmap_remove_and_put() or remove_and_replace()?

Reply:
(yes i know the old code didn't use that either, but let's fix that while we are at it)

Reply:
I guess you suggesting upsert i.e.  do the operation in a single call, is that right?

I don't see a way to do that.  The prior value (if any) needs to be freed but ordered_hashmap_remove_and_put() + remove_and_replace() don't free it nor return it.  Also, both of those routines require the old key to be present, and it might not be present here.

Given the current interface it could be also be expressed like this which I think is at least preferable to what was there before, so I have done that:

```
        old = ordered_hashmap_get(*options, UINT_TO_PTR(u));
        r = ordered_hashmap_replace(*options, UINT_TO_PTR(u), opt);
```

Please let me know.  Thank you!